demos/fleet_provisioning/fleet_provisioning_with_csr/fleet_provisioning_with_csr_demo.c

```diff
@@ -0,0 +1,883 @@
+
+/**
+ * @brief Obtains the CSR from the file path in #PROVISIONING_CSR_PATH.
+ *
+ * @param[in] pBuffer Buffer into which to read the CSR.
+ * @param[in] bufferLength Length of #buffer.
+ * @param[out] pOutCsrLength The length of the CSR.
+ */
+static bool getCsr( char * pBuffer,
+                    size_t bufferLength,
+                    size_t * pOutCsrLength );
+
+/**
+ * @brief Run the MQTT process loop to get a response.
+ */
+static bool waitForResponse( void );
+
+/**
+ * @brief Subscribe to the CreateCertificateFromCsr accepted and rejected topics.
+ */
+static bool subscribeToCsrResponseTopics( void );
+
+/**
+ * @brief Unsubscribe from the CreateCertificateFromCsr accepted and rejected topics.
+ */
+static bool unsubscribeFromCsrResponseTopics( void );
+
+/**
+ * @brief Subscribe to the RegisterThing accepted and rejected topics.
+ */
+static bool subscribeToRegisterThingResponseTopics( void );
+
+/**
+ * @brief Unsubscribe from the RegisterThing accepted and rejected topics.
+ */
+static bool unsubscribeFromRegisterThingResponseTopics( void );
+
+/**
+ * @brief Save the certificate to the path specified by #PROVISIONING_CERT_PATH.
+ *
+ * @param[in] pCertificate The certificate to save.
+ * @param[in] certificateLength Length of #pCertificate.
+ */
+static bool saveCertificate( const char * pCertificate,
+                             size_t certificateLength );
+/*-----------------------------------------------------------*/
+
+static void provisioningPublishCallback( MQTTPublishInfo_t * pPublishInfo,
+                                         uint16_t packetIdentifier )
+{
+    FleetProvisioningStatus_t status;
+    FleetProvisioningTopic_t api;
+
+    /* Silence compiler warnings about unused variables. */
+    ( void ) packetIdentifier;
+
+    status = FleetProvisioning_MatchTopic( pPublishInfo->pTopicName,
+                                           pPublishInfo->topicNameLength, &api );
+
+    if( status != FleetProvisioningSuccess )
+    {
+        LogWarn( ( "Unexpected publish message received. Topic: %.*s.",
+                   ( int ) pPublishInfo->topicNameLength,
+                   ( const char * ) pPublishInfo->pTopicName ) );
+    }
+    else
+    {
+        if( api == FleetProvCborCreateCertFromCsrAccepted )
+        {
+            LogInfo( ( "Received accepted response from Fleet Provisioning CreateCertificateFromCsr API." ) );
+
+            responseStatus = ResponseAccepted;
+
+            /* Copy the payload from the MQTT library's buffer to #payloadBuffer. */
+            ( void ) memcpy( ( void * ) payloadBuffer,
+                             ( const void * ) pPublishInfo->pPayload,
+                             ( size_t ) pPublishInfo->payloadLength );
+
+            payloadLength = pPublishInfo->payloadLength;
+        }
+        else if( api == FleetProvCborCreateCertFromCsrRejected )
+        {
+            LogError( ( "Received rejected response from Fleet Provisioning CreateCertificateFromCsr API." ) );
```

Comment:
How about parsing the error response payload provided by IoT Core? `CreateCertificateFromCsr` provides cause of error information in the response payload as :

```
{
    "statusCode": int,
    "errorCode": "string",
    "errorMessage": "string"
}
```

Similarly, for the `RegisterThing` rejected response

Reply:
I'll leave that for follow up PR with printing the CBOR info